Feature/push package collector

janbritz opened the pull request:
Closes #23, closes #24.

Die einzelnen Kollektoren des `PackageCollector`s updaten den `Indexer` bei Veränderungen ihres Paketbestandes nun selbst.
Außerdem initialisiert jetzt der `Indexer` die `Package`s, statt wie bisher die Sub-Kollektoren.
Existiert beim Registrieren eines Pakets bereits die dazugehörige `Package`-Klasse im Indexer, so wird der Kollektor in den `PackageSources` vermerkt. Dadurch wird nur eine `Package`-Klasse erstellt, selbst wenn ein Paket von mehreren Kollektoren angeboten wird
Anhand dieser `PackageSources` entscheidet der `Indexer` über die Visibilität des Pakets.

Der `LocalCollector` reagiert nun bei jeglichen Änderungen (Umbenennung, Modifizierung, Löschung, Erstellung) der Pakete im zugehörigen Ordner, updatet seine interne `LocalCollector.map` und nimmt ggf. auch Änderungen am `Indexer` vor.

Zusätzlich wurde das Pylint-Plugin [pylint-pytest](https://github.com/reverbc/pylint-pytest) entfernt, da dieses Bugs enthält und archiviert wurde. Dadurch werden nun beim Ausführen des Linters, zuvor ignorierte Warnungen (`redefined-outer-name` und `protected-access`) ausgegeben, die unseren Testcode betreffen.
Es ist leider ([noch](https://github.com/PyCQA/pylint/issues/618)) nicht möglich, bestimmte Warnungen pro Ordner zu ignorieren, allerdings könnte man entweder den gesamten `questionpy-server/tests/`-Ordner ignorieren, `# pylint: disable=...` an den Anfang jeder Test-Datei kopieren oder Pylint mit einer anderen Konfiguration auf den Test-Ordner laufen lassen.

janbritz committed:
feat: `PackageCollector` pushes packages to `Indexer`

The sub-collectors (currently only `LMSCollector` and `LocalCollector`) push available packages to the `Indexer` and notify it on the removal or modification of a package.

janbritz committed:
chore: update dependencies

MartinGauk commented on questionpy_server/collector/indexer.py:
```diff
@@ -59,73 +53,87 @@ async def get_by_name_and_version(self, short_name: str, version: str) -> Option
+    async def register_package(self, package_hash: str, path_or_manifest: Union[Path, Manifest],
+                               source: BaseCollector) -> Package:
+
+        if package := self._index_by_hash.get(package_hash, None):
```
Du brauchst für diese Funktion einen async lock. Wenn die Funktion mit demselben Paket (path) mehrmals gleichzeitig aufgerufen wird, kommen die sich in die Quere ansonsten.

MartinGauk commented on questionpy_server/collector/package_collector.py:
```diff
@@ -0,0 +1,112 @@
+    from questionpy_server.package import Package
+
+
+class PackageCollector:
```
Ich würde vorschlagen, dass wir die Klasse (und den Dateinamen) umbenennen in `PackageCollection`, um die Klasse namentlich stärker von den eigentlichen Collectors abzugrenzen.

MartinGauk commented on questionpy_server/collector/package_collector.py:
```diff
@@ -0,0 +1,112 @@
+                continue
+            # We assume that existing packages from the cache are from an LMS.
+            coroutine = self._indexer.register_package(calculate_hash(file), file, self._lms_collector)
+            run(coroutine)
```
Die Initialisierung findet gerade doppelt statt (hier und in `LMSCollector.__init__`). Ich glaube `async start/stop` Funktionen für jeden collector und für diese Klasse wären sinnvoll. Dann könnte der LocalCollector.observer auch gestoppt werden beim Beenden des Servers, was sauberer ist.
Diese Klasse sollte dann gar keine Pakete laden, sondern einfach nur die start Methoden der Collector aufrufen. Die könnten dann sogar mittels `asyncio.gather` gleichzeitig aufgerufen werden.

MartinGauk replied:
Sry, ich kam da vorhin durcheinander. Die Initialisierung findet nicht doppelt statt, allerdings fände ich eine Verschiebung des Codes sinnvoll. Außerdem könntest du im LMSCollector auf die Cache-Klasse zurückgreifen, die den Cache-Ordner ja bereits eingelesen hat. Der Hash steckt ja zudem schon im Dateinamen und muss nicht neu berechnet werden.

MartinGauk commented on questionpy_server/collector/local_collector.py:
```diff
@@ -0,0 +1,209 @@
+        await self._local_collector.indexer.register_package(package_hash, path, self._local_collector)
+
+    def _push_package(self, path_str: str) -> None:
+        run(self._push_package_async(path_str))
```
Die Event-Handler werden offenbar in einem anderen Thread ausgeführt. Du musst daher mittels `asyncio.run_coroutine_threadsafe` die Coroutine in unserer main loop ausführen lassen.

Die `_pop_package` Methode darf ebenfalls nicht einfach so aus einem anderen Thread im Indexer Änderungen machen. Verwende dort am besten auch `asyncio.run_coroutine_threadsafe`.

janbritz committed:
refactor: rename `PackageCollector` to `PackageCollection`

janbritz committed:
feat: add `start()` and `stop()` to collectors

janbritz committed:
tests: adding test cases

janbritz committed:
test: use `patch.object()` instead of `patch()`

TODO: Replace every `patch()` with `patch.object()` when possible..

MartinGauk commented on questionpy_server/collector/indexer.py:
```diff
@@ -86,25 +90,33 @@ async def register_package(self, package_hash: str, path_or_manifest: Path, sour
+                if existing_package and existing_package != package:
+                    # Package with the same short_name and version already exists; log a warning.
+                    log = logging.getLogger('questionpy-server')
+                    log.warning("The package %s (%s) already exists with a different hash.",
```
Hier bitte auch die zwei Hashes ausgeben.

MartinGauk commented on questionpy_server/collector/lms_collector.py:
```diff
@@ -20,8 +21,21 @@ class LMSCollector(CachedCollector):
+        count = 0
+        # We assume that existing packages in the cache are from an LMS as it has the most strict visibility i.e. the
+        # package can only be accessed by the hash.
+        for file in self._cache.directory.iterdir():
```
Du musst nicht auf das Dateisystem zugreifen, wenn das Cache-Objekt doch auch schon alle Dateien geladen hat.

MartinGauk commented on questionpy_server/collector/local_collector.py:
```diff
@@ -172,38 +183,48 @@ class LocalCollector(BaseCollector):
-        self.observer.schedule(event_handler, str(directory))
-        self.observer.start()
+        # Start the directory observer.
+        self._event_handler = LocalCollectorEventHandler(self, get_event_loop())
```
get_event_loop ist deprecated. Nimm lieber `get_running_loop()`.

MartinGauk commented on questionpy_server/collector/local_collector.py:
```diff
@@ -114,24 +115,28 @@ class LocalCollectorEventHandler(PatternMatchingEventHandler):
+
         self._local_collector = local_collector
+        self._loop = loop
+        self._log = logging.getLogger('questionpy-server')
```
Benutze am besten innerhalb dieses Moduls nur dasselbe Log-Objekt und benenne es `questionpy-server-local-collector` oder so.

MartinGauk commented on questionpy_server/collector/local_collector.py:
```diff
@@ -114,24 +115,28 @@ class LocalCollectorEventHandler(PatternMatchingEventHandler):
+        ).result()
 
-    def _pop_package(self, path_str: str) -> None:
+    def _remove_package(self, path_str: str) -> None:
```
Beachte hier bitte weiterhin meinen anderen Kommentar. Du darfst nicht einfach aus einem anderen Thread im Indexer Änderungen machen.

janbritz committed:
fix: Lock was not initialized in the right eventloop

The `asyncio.Lock` needs to be initialized inside the webservers eventloop. Otherwise, this causes a `RuntimeError`. This error does not occur for the Python versions 3.10 and above.

janbritz committed:
test: Use `patch.object()` instead of `patch()`

janbritz committed:
feat: Make `Indexer.unregister_package()` async

janbritz committed:
feat: `Cache.files` returns every file in the cache

janbritz committed:
refactor: Make logging source clearer

janbritz committed:
chore: change order of directories in pylint call

janbritz committed:
chore: switch arguments in pylint call

janbritz committed:
chore: remove pylint-pytest plugin

janbritz committed:
test: test if file changes get registered
